tlscommon.c

```diff
@@ -626,13 +628,182 @@ static int certattr_matchip(GENERAL_NAME *gn, struct certattrmatch *match){
+    char* firstPorCertDNS = extractFirstPortion(certDNS, '.');
+    if(firstPorCertDNS == NULL) {
+        return 0;
+    }
+    char* firstPorHostName = extractFirstPortion(hostname, '.');
+    if(firstPorHostName == NULL) {
+        free(firstPorCertDNS);
+        return 0;
+    }
+    
+    char* pattern = NULL;
+    char* regex2 = "[a-zA-Z0-9-]";
+    char* startRegex = "^";
+    char* endRregex = "$";
+    char* wildcard = "*";
+    
+    // Handles *.3af521.net (cert DNS) and idp.3af521.net (host name) case
+    if(strlen(firstPorCertDNS) == 1 && firstPorCertDNS[0] == '*') {
+
+        char* lastPorHostName = strchr(hostname, '.');
+        if(lastPorHostName == NULL) {
+            free(firstPorCertDNS);
+            free(firstPorHostName);
+            return 0;
+        }
+        char* modifiedHostName = append(2, wildcard, lastPorHostName);
+        if(modifiedHostName == NULL) {
+            free(firstPorCertDNS);
+            free(firstPorHostName);
+            return 0;
+        }
+        if(strlen(modifiedHostName) == strlen(certDNS) && memcmp(modifiedHostName, certDNS, strlen(certDNS)) == 0) {
+            free(modifiedHostName);
+            free(firstPorCertDNS);
+            free(firstPorHostName);
+            return 1;
+        }
+        free(modifiedHostName);
+        free(firstPorCertDNS);
+        free(firstPorHostName);
+        return 0;
+    }
+    
+    char* asteriskPos = strchr(firstPorCertDNS, '*');
+    if(asteriskPos == NULL) {
+        free(firstPorCertDNS);
+        free(firstPorHostName);
+        return 0;
+    }
+
+    char* lenRegex = calculateLenOfLabel(63 - strlen(firstPorCertDNS) - 1);
+    if(lenRegex == NULL) {
+        free(firstPorCertDNS);
+        free(firstPorHostName);
+        return 0;
+    }
+
+    if(firstPorCertDNS[0] == '*') {
+        char* regex1 = "^[a-z0-9A-Z][a-z0-9A-Z-]";
+        pattern = append(4, regex1, lenRegex, asteriskPos + 1, endRregex);
+    } else if(firstPorCertDNS[strlen(firstPorCertDNS)-1] == '*') {
+        char* regex1 = extractFirstPortion(firstPorCertDNS, '*');
+        pattern = append(5, startRegex, regex1, regex2, lenRegex, endRregex);
+    } else {
+        char* regex1 = extractFirstPortion(firstPorCertDNS, '*');
+        pattern = append(6, startRegex, regex1, regex2, lenRegex, asteriskPos + 1, endRregex);
+    }
+    
+    if(regexMatcher(pattern, firstPorHostName)) {
+        free(pattern);
+        free(firstPorCertDNS);
+        free(firstPorHostName);
+        return 1;
+    }
+    free(pattern);
+    free(firstPorCertDNS);
+    free(firstPorHostName);
+    return 0;
+}
+
 static int _general_name_regex_match(char *v, int l, struct certattrmatch *match) {
     char *s;
     if (l <= 0 ) 
         return 0;
     if (match->exact) {
         if (l == strlen(match->exact) && memcmp(v, match->exact, l) == 0)
             return 1;
+        if(compareWithModifiedHostname(v, match->exact)) {
```

Comment:
Why is this in the `match->exact` block?  If it must be here, then please include a comment on why since it mutates the expected behaviour of an `exact` match.

Reply:
This is the original code format. If match->exact is set it means we have found a host/DNS (the script has returned a valid host).